### ISSUE TITLE
Newbie documentation for Chef.set_resource_priority_array

### ISSUE THREAD

**Issue by lonniev** [chef/chef#3349]
Chef 12.3.0 introduces dynamic resolution of resources when they are short-named in recipes. The release page claims that users can simply add `Chef.set_resource_priority_array` statements to library files so that the cookbook recipes compile properly. One example for MacPorts is given.

A complete example showing the entire format of the library file and showing the inclusion of that library into the recipe file would be very useful (to me at least).

I find that regardless of the hacked syntax I attempt, the necessary recipe continues to complain about the ambiguity of its resource use. (For me, the particular resource is `windows_package` and the Windows Cookbook.)

**Comment by lonniev** [comment]
I add a libraries/ambiguity_resolver.rb file that has only the following:

```ruby
Chef.set_resource_priority_array(:windows_package, [ Chef::Resource::WindowsCookbookPackage ], os: :windows)

module Ambiguity
  module Resolver
  end
end
```

In the cookbook/recipes/default.rb, I attempt to include that crucial `Chef.set_resource_priority_array` statement with this statement:

```ruby
::Chef::Recipe.send(:include, Ambiguity::Resolver)
```

When provisioning from vagrant, I continue to receive this warning:

```bash
==> vagrant-machine-name: [2015-05-07T09:20:46-07:00] WARN: Ambiguous resource precedence: [Chef::Resource::WindowsCookbookPackage, Chef::Resource::WindowsPackage], please use Chef.set_resource_priority_array to provide determinism
```

**Comment by lonniev** [comment]
Perhaps it is only a typo in the Chef Release notes where the `filter` argument was given? When I switch from `os: :windows` or `os: :windows` to `platform: "windows"` then the warning goes away, suggesting that the library was able to tell Chef how to resolve the ambiguous references.

**Comment by lonniev** [comment]
In case some other Windows Cookbook user runs into this new feature of Chef 12.3.0, here is the library file that makes the warning go away.

```ruby
Chef.set_resource_priority_array(:windows_package, [ Chef::Resource::WindowsCookbookPackage ], platform: "windows")

module Ambiguity
  module Resolver
  end
end
```

**Comment by lamont-granquist** [comment]
You only need to include the Chef.set_resource_priority_array one-liner, you don't need the module definition.  Other than that, you've got it correct.

**Comment by lonniev** [comment]
Yes, the reason for the module def is to gain a namespace that is indicative and by which I can include the library by indicative means. Without the namespace, I can include the library but I can’t express why in the client code—or so I believe.




My current challenge is that the actual recipe uses BOTH windows_package from pure Chef and from WindowsCookbook. Is there a way to say use one provider for one instance of an LWRP and use a different provider for another instance in the same recipe?

**Comment by lamont-granquist** [comment]
libraries are require'd by chef-client automatically, you don't need to require it or reference it, that code will also get evaluated when the 'class' is compiled (sorta like class loading time code, but not really at ruby class loading time, but chef-client library loading time) and therefore you don't need to poke it in any way from recipe code (no need to include/require/extend or anything).

**Comment by lamont-granquist** [comment]
And right now i think you can't do that because the code in the cookbook takes the cookbook version of windows_package and deliberately scribbles over the internal version of cookbook_package so you can't use both of them.  I'm submitting a PR to the windows cookbook right now to allow that kind of thing.

**Comment by lamont-granquist** [comment]
Afterwards you'll need to do something like

```
Chef::Resource::WindowsPackage.provides :windows_package_core, os: "windows"
```

Then `windows_package_core` resource in the DSL would map to the internal one.  But I suspect if you try that now that you'll still wind up with the cookbook one.

**Comment by lonniev** [comment]
Isn't there a "provides" attribute of every resource (LWRP) DSL instance (i.e. for each "call" of the resource) where the recipe author can state which cookbook provider they intend to resolve the resource reference?

Maybe I should throw away the attempt to keep the dynamic resolver happy and just state exactly which one resource I want to use?

**Comment by lamont-granquist** [comment]
There's a provider parameter which you can use:

```ruby
windows_package "whatever" do
  provider Chef::Provider::Package::Windows 
end
```

What you wind up with when you do that, however, is a windows_package resource from the cookbook wrapping the core windows_package provider.  You don't get a core windows_package resource wrapping a core windows_package provider.  If the resources have differences in the shape of their API implementation, then you might find that doesn't work (although if it works, then that is by far the easiest thing to do).

**Comment by adamedx** [comment]
@jkeiser, is there documentation for the 12.4 and 12.5 resource changes?

**Comment by jkeiser** [comment]
In the DOC_CHANGES, yep! I'm not sure where those go when a release happens, though ...

**Comment by lamont-granquist** [comment]
Yeah, at some point we need a blog post about the way to do Chef >= 12.4.0 provider priority, and the way to do Chef 11.x/12.x back-compat provider priority.

I don't think we have this warning any more and the way to fix it in Chef >= 12.4.0 doesn't involve poking the map directly but just adding more provides lines typically.

**Comment by lamont-granquist** [comment]
We shouldn't document this, if anything we should deprecate it.
